File: authorization-for-rag/langgraph-state-js/helpers/fga-retriever.ts
import { Document, DocumentInterface } from "@langchain/core/documents";
import { BaseRetriever } from "@langchain/core/retrievers";
import {
  ClientBatchCheckItem,
  ConsistencyPreference,
  CredentialsMethod,
  OpenFgaClient,
} from "@openfga/sdk";

import type { BaseRetrieverInput } from "@langchain/core/retrievers";
import type { CallbackManagerForRetrieverRun } from "@langchain/core/callbacks/manager";

export type FGARetrieverCheckerFn = (
  doc: DocumentInterface<Record<string, any>>
) => ClientBatchCheckItem;

export type FGARetrieverArgs = {
  retriever: BaseRetriever;
  buildQuery: FGARetrieverCheckerFn;
  fields?: BaseRetrieverInput;
};

type AccessByDocumentFn = (
  checks: ClientBatchCheckItem[]
) => Promise<Map<string, boolean>>;

type FGARetrieverArgsWithAccessByDocument = FGARetrieverArgs & {
  accessByDocument: AccessByDocumentFn;
};

/**
 * A retriever that allows filtering documents based on access control checks
 * using OpenFGA. This class wraps an underlying retriever and performs batch
 * checks on retrieved documents, returning only the ones that pass the
 * specified access criteria.
 *
 * @remarks
 * The FGARetriever requires a buildQuery function to specify how access checks
 * are formed for each document, the checks are executed via an OpenFGA client
 * or equivalent mechanism. The checks are then mapped back to their corresponding
 * documents to filter out those for which access is denied.
 *
 * @example
 * ```ts
 * const retriever = FGARetriever.create({
 *   retriever: someOtherRetriever,
 *   buildQuery: (doc) => ({
 *     user: `user:${user}`,
 *     object: `doc:${doc.metadata.id}`,
 *     relation: "viewer",
 *   }),
 * });
 * ```
 */
export class FGARetriever extends BaseRetriever {
  lc_namespace = ["@langchain", "retrievers"];
  private retriever: BaseRetriever;
  private buildQuery: FGARetrieverCheckerFn;
  private accessByDocument: AccessByDocumentFn;

  private constructor({
    retriever,
    buildQuery,
    fields,
    accessByDocument,
  }: FGARetrieverArgsWithAccessByDocument) {
    super(fields);
    this.buildQuery = buildQuery;
    this.retriever = retriever;
    this.accessByDocument = accessByDocument as AccessByDocumentFn;
  }

  /**
   * Creates a new FGARetriever instance using the given arguments and optional OpenFgaClient.
   *
   * @param args - @FGARetrieverArgs
   * @param args.retriever - The underlying retriever instance to fetch documents.
   * @param args.buildQuery - A function to generate access check requests for each document.
   * @param args.fields - Optional - Additional fields to pass to the underlying retriever.
   * @param fgaClient - Optional - OpenFgaClient instance to execute checks against.
   * @returns A newly created FGARetriever instance configured with the provided arguments.
   */
  static create(
    args: FGARetrieverArgs,
    fgaClient?: OpenFgaClient
  ): FGARetriever {
    const client =
      fgaClient ||
      new OpenFgaClient({
        apiUrl: process.env.FGA_API_URL || "https://api.us1.fga.dev",
        storeId: process.env.FGA_STORE_ID!,
        credentials: {
          method: CredentialsMethod.ClientCredentials,
          config: {
            apiTokenIssuer: process.env.FGA_API_TOKEN_ISSUER || "auth.fga.dev",
            apiAudience:
              process.env.FGA_API_AUDIENCE || "https://api.us1.fga.dev/",
            clientId: process.env.FGA_CLIENT_ID!,
            clientSecret: process.env.FGA_CLIENT_SECRET!,
          },
        },
      });

    const accessByDocument: AccessByDocumentFn = async function (checks) {
      const response = await client.batchCheck(
        { checks },
        {
          consistency: ConsistencyPreference.HigherConsistency,
        }
      );
      return response.result.reduce(
        (permissionMap: Map<string, boolean>, result) => {
          permissionMap.set(result.request.object, result.allowed || false);
          return permissionMap;
        },
        new Map<string, boolean>()
      );
    };

    return new FGARetriever({ ...args, accessByDocument });
  }

  /**
   * Retrieves documents based on the provided query parameters, processes
   * them through a checker function,
   * and filters the documents based on permissions.
   *
   * @param params - The query parameters used to retrieve nodes.
   * @returns A promise that resolves to an array of documents that have passed the permission checks.
   */
  async _getRelevantDocuments(
    query: string,
    runManager?: CallbackManagerForRetrieverRun
  ): Promise<Document[]> {
    const documents = await this.retriever._getRelevantDocuments(
      query,
      runManager
    );

    const { checks, documentToObject } = documents.reduce(
      (acc, doc) => {
        const check = this.buildQuery(doc);
        acc.checks.push(check);
        acc.documentToObject.set(doc, check.object);
        return acc;
      },
      {
        checks: [] as ClientBatchCheckItem[],
        documentToObject: new Map<
          DocumentInterface<Record<string, any>>,
          string
        >(),
      }
    );

    const resultsByObject = await this.accessByDocument(checks);

    return documents.filter(
      (d, i) => resultsByObject.get(documentToObject.get(d) || "") === true
    );
  }
}

File: authorization-for-rag/langchain4j-java/src/main/java/rag/FGARetriever.java
package rag;

import dev.langchain4j.rag.content.Content;
import dev.langchain4j.rag.content.retriever.ContentRetriever;
import dev.langchain4j.rag.query.Query;
import dev.openfga.sdk.api.client.OpenFgaClient;
import dev.openfga.sdk.api.client.model.ClientBatchCheckItem;
import dev.openfga.sdk.api.client.model.ClientBatchCheckRequest;
import dev.openfga.sdk.api.configuration.ClientBatchCheckOptions;
import dev.openfga.sdk.api.configuration.ClientConfiguration;
import dev.openfga.sdk.api.configuration.ClientCredentials;
import dev.openfga.sdk.api.configuration.Credentials;
import dev.openfga.sdk.api.model.ConsistencyPreference;
import dev.openfga.sdk.errors.FgaInvalidParameterException;
import dev.openfga.sdk.errors.FgaValidationError;

import java.util.*;
import java.util.concurrent.ExecutionException;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * A retriever that allows filtering documents based on access control checks
 * using OpenFGA. This class wraps an underlying retriever and performs batch
 * checks on retrieved documents, returning only the ones that pass the
 * specified access criteria.
 *
 * <p>
 * The FGARetriever requires a buildQuery function to specify how access checks
 * are formed for each document, the checks are executed via an OpenFGA client
 * or equivalent mechanism. The checks are then mapped back to their
 * corresponding
 * documents to filter out those for which access is denied.
 * </p>
 *
 * <p>
 * Example usage:
 * </p>
 *
 * <pre>
 * FGARetriever retriever = FGARetriever.create(
 *         baseRetriever,
 *         "user1",
 *         content -> new FGATuple(
 *                 "user:" + userId,
 *                 "viewer",
 *                 "doc:" + content.metadata().get("id")));
 * </pre>
 */
public class FGARetriever implements ContentRetriever {
    private final ContentRetriever baseRetriever;
    private final OpenFgaClient fgaClient;
    private final ConsistencyPreference consistencyPreference;
    private final Function<Content, ClientBatchCheckItem> buildQuery;

    FGARetriever(ContentRetriever baseRetriever, Function<Content, ClientBatchCheckItem> buildQuery, ConsistencyPreference consistencyPreference, OpenFgaClient fgaClient) throws Exception {
        this.baseRetriever = baseRetriever;
        this.fgaClient = fgaClient;
        this.buildQuery = buildQuery;
        this.consistencyPreference = consistencyPreference;
    }

    FGARetriever(ContentRetriever baseRetriever, Function<Content, ClientBatchCheckItem> buildQuery, ConsistencyPreference consistencyPreference) throws Exception {
        this(baseRetriever,
                buildQuery,
                consistencyPreference,
                new OpenFgaClient(new ClientConfiguration()
                        .apiUrl(System.getProperty("FGA_API_URL", "https://api.us1.fga.dev"))
                        .storeId(System.getProperty("FGA_STORE_ID"))
                        .credentials(new Credentials(new ClientCredentials()
                                .apiTokenIssuer(System.getProperty("FGA_API_TOKEN_ISSUER", "auth.fga.dev"))
                                .apiAudience(System.getProperty("FGA_API_AUDIENCE", "https://api.us1.fga.dev/"))
                                .clientId(System.getProperty("FGA_CLIENT_ID"))
                                .clientSecret(System.getProperty("FGA_CLIENT_SECRET"))
                        ))
                )
        );

    }

    FGARetriever(ContentRetriever baseRetriever, Function<Content, ClientBatchCheckItem> buildQuery) throws Exception {
        this(baseRetriever, buildQuery, ConsistencyPreference.HIGHER_CONSISTENCY);
    }

    /**
     * Creates a new FGARetriever instance using the given arguments and optional OpenFgaClient.
     *
     * @param retriever  - The underlying retriever instance to fetch documents.
     * @param buildQuery - A function to generate access check requests for each document.
     * @return A newly created FGARetriever instance configured with the provided arguments.
     */
    public static FGARetriever create(ContentRetriever retriever, Function<Content, ClientBatchCheckItem> buildQuery) throws Exception {
        return new FGARetriever(retriever, buildQuery);
    }


    /**
     * Retrieves documents based on the provided query parameters, processes
     * them through a checker function,
     * and filters the documents based on permissions.
     *
     * @param query - The query parameters used to retrieve content.
     * @return Filtered Content
     */
    @Override
    public List<Content> retrieve(Query query) {
        // First, get relevant documents from the base retriever
        List<Content> relevantContent = baseRetriever.retrieve(query);

        // Create data structures to track checks and document mappings
        List<ClientBatchCheckItem> checks = new ArrayList<>();
        var documentToObject = new HashMap<>();
        var seenChecks = new HashSet<>();

        // Process each document to build checks
        for (Content doc : relevantContent) {
            ClientBatchCheckItem check = buildQuery.apply(doc);
            var checkKey = getCheckKey(check);
            documentToObject.put(doc, checkKey);

            // Skip duplicate checks for same user, object, and relation
            if (!seenChecks.contains(checkKey)) {
                seenChecks.add(checkKey);
                checks.add(check);
            }
        }

        try {
            var permissionsMap = checkPermissions(checks);
            // filter based on permission
            return relevantContent.stream()
                    .filter(doc -> Boolean.TRUE.equals(permissionsMap.get(documentToObject.get(doc))))
                    .collect(Collectors.toList());
        } catch (FgaInvalidParameterException | FgaValidationError | ExecutionException | InterruptedException e) {
            throw new RuntimeException(e);
        }


    }

    /**
     * Checks permissions for a list of client requests.
     *
     * @param checks - An array of `ClientBatchCheckItem` objects representing the permissions to be checked.
     * @return A `Map` where the keys are object identifiers and the values are booleans indicating whether the permission is allowed.
     */
    private Map<String, Boolean> checkPermissions(List<ClientBatchCheckItem> checks) throws FgaInvalidParameterException, FgaValidationError, ExecutionException, InterruptedException {
        var options = new ClientBatchCheckOptions().consistency(consistencyPreference);
        var request = new ClientBatchCheckRequest().checks(checks);
        var response = fgaClient.batchCheck(request, options).get();

        Map<String, Boolean> permissionMap = new HashMap<>();
        for (var result : response.getResult()) {
            var checkKey = getCheckKey(result.getRequest());
            permissionMap.put(checkKey, result.isAllowed());
        }
        return permissionMap;
    }

    private String getCheckKey(ClientBatchCheckItem check) {
        return check.getUser() + "|" + check.getObject() + "|" + check.getRelation();
    }
}